File: go/pkg/memory/compressor.go
package memory

import (
	"context"
	"fmt"
	"strings"

	"github.com/haivivi/giztoy/go/pkg/genx/profilers"
	"github.com/haivivi/giztoy/go/pkg/genx/segmentors"
)

// LLMCompressor implements [Compressor] by delegating to the segmentors and
// profilers packages for LLM-based conversation compression.
//
// It calls [segmentors.Process] to extract entities, relations, and a
// compressed segment summary. Optionally, it calls [profilers.Process] to
// evolve entity schemas and update profiles.
//
// LLMCompressor is stateless and safe for concurrent use.
type LLMCompressor struct {
	cfg LLMCompressorConfig

	segMux  *segmentors.Mux
	profMux *profilers.Mux
}

// LLMCompressorConfig configures an [LLMCompressor].
type LLMCompressorConfig struct {
	// Segmentor is the pattern of the registered segmentor to use
	// (e.g., "seg/qwen-flash"). Required. Must be registered in
	// segmentors.DefaultMux (or the provided SegmentorMux).
	Segmentor string

	// Profiler is the pattern of the registered profiler to use
	// (e.g., "prof/qwen-flash"). Optional. If empty, profiling is
	// skipped and only segmentor output is used.
	Profiler string

	// Schema provides entity type hints to guide extraction.
	// Optional. If nil, the LLM discovers entity types freely.
	Schema *segmentors.Schema

	// Profiles holds current entity profiles for the profiler to
	// reference when proposing updates. Optional. Keyed by entity
	// label (e.g., "person:小明") → attribute map.
	Profiles map[string]map[string]any

	// SegmentorMux overrides segmentors.DefaultMux. Optional.
	SegmentorMux *segmentors.Mux

	// ProfilerMux overrides profilers.DefaultMux. Optional.
	ProfilerMux *profilers.Mux
}

// NewLLMCompressor creates a new LLM-based compressor.
func NewLLMCompressor(cfg LLMCompressorConfig) (*LLMCompressor, error) {
	if cfg.Segmentor == "" {
		return nil, fmt.Errorf("memory: LLMCompressorConfig.Segmentor is required")
	}
	return &LLMCompressor{
		cfg:     cfg,
		segMux:  cfg.SegmentorMux,
		profMux: cfg.ProfilerMux,
	}, nil
}

// CompressMessages compresses conversation messages into memory segments
// and an updated summary by calling the segmentor LLM.
func (c *LLMCompressor) CompressMessages(ctx context.Context, messages []Message) (*CompressResult, error) {
	result, err := c.runSegmentor(ctx, messages)
	if err != nil {
		return nil, err
	}

	// Convert segmentor output to CompressResult.
	seg := SegmentInput{
		Summary:  result.Segment.Summary,
		Keywords: result.Segment.Keywords,
		Labels:   result.Segment.Labels,
	}

	return &CompressResult{
		Segments: []SegmentInput{seg},
		Summary:  result.Segment.Summary,
	}, nil
}

// ExtractEntities extracts entity and relation updates from messages
// by calling the segmentor (and optionally the profiler) LLM.
func (c *LLMCompressor) ExtractEntities(ctx context.Context, messages []Message) (*EntityUpdate, error) {
	result, err := c.runSegmentor(ctx, messages)
	if err != nil {
		return nil, err
	}

	update := convertSegmentorResult(result)

	// Run profiler if configured.
	if c.cfg.Profiler != "" {
		profResult, err := c.runProfiler(ctx, messages, result)
		if err != nil {
			// Profiler failure is non-fatal — we still have the segmentor result.
			// Log and continue (the caller may decide to log this).
			_ = err
		} else {
			mergeProfilerResult(update, profResult)
		}
	}

	return update, nil
}

// runSegmentor calls segmentors.Process with the converted messages.
func (c *LLMCompressor) runSegmentor(ctx context.Context, messages []Message) (*segmentors.Result, error) {
	input := segmentors.Input{
		Messages: messagesToStrings(messages),
		Schema:   c.cfg.Schema,
	}

	if c.segMux != nil {
		return c.segMux.Process(ctx, c.cfg.Segmentor, input)
	}
	return segmentors.Process(ctx, c.cfg.Segmentor, input)
}

// runProfiler calls profilers.Process with the segmentor result and context.
func (c *LLMCompressor) runProfiler(ctx context.Context, messages []Message, segResult *segmentors.Result) (*profilers.Result, error) {
	input := profilers.Input{
		Messages:  messagesToStrings(messages),
		Extracted: segResult,
		Schema:    c.cfg.Schema,
		Profiles:  c.cfg.Profiles,
	}

	if c.profMux != nil {
		return c.profMux.Process(ctx, c.cfg.Profiler, input)
	}
	return profilers.Process(ctx, c.cfg.Profiler, input)
}

// messagesToStrings converts memory.Message slice to the plain string format
// expected by segmentors/profilers: "role: content".
func messagesToStrings(messages []Message) []string {
	out := make([]string, 0, len(messages))
	for _, m := range messages {
		if m.Content == "" {
			continue
		}
		var sb strings.Builder
		sb.WriteString(string(m.Role))
		if m.Name != "" {
			sb.WriteByte('(')
			sb.WriteString(m.Name)
			sb.WriteByte(')')
		}
		sb.WriteString(": ")
		sb.WriteString(m.Content)
		out = append(out, sb.String())
	}
	return out
}

// convertSegmentorResult converts segmentor entities and relations into
// a memory.EntityUpdate.
func convertSegmentorResult(result *segmentors.Result) *EntityUpdate {
	update := &EntityUpdate{}

	for _, e := range result.Entities {
		update.Entities = append(update.Entities, EntityInput{
			Label: e.Label,
			Attrs: e.Attrs,
		})
	}

	for _, r := range result.Relations {
		update.Relations = append(update.Relations, RelationInput{
			From:    r.From,
			To:      r.To,
			RelType: r.RelType,
		})
	}

	return update
}

// mergeProfilerResult merges profiler output into an existing EntityUpdate.
// Profile updates are merged as entity attrs; additional relations are appended.
func mergeProfilerResult(update *EntityUpdate, profResult *profilers.Result) {
	if profResult == nil {
		return
	}

	// Merge profile updates as entity attrs.
	for label, attrs := range profResult.ProfileUpdates {
		found := false
		for i, e := range update.Entities {
			if e.Label == label {
				// Merge attrs into existing entity.
				if update.Entities[i].Attrs == nil {
					update.Entities[i].Attrs = make(map[string]any)
				}
				for k, v := range attrs {
					update.Entities[i].Attrs[k] = v
				}
				found = true
				break
			}
		}
		if !found {
			update.Entities = append(update.Entities, EntityInput{
				Label: label,
				Attrs: attrs,
			})
		}
	}

	// Append additional relations from profiler.
	for _, r := range profResult.Relations {
		update.Relations = append(update.Relations, RelationInput{
			From:    r.From,
			To:      r.To,
			RelType: r.RelType,
		})
	}
}

File: rust/memory/src/compressor.rs
//! LLMCompressor - LLM-based conversation compressor
//!
//! This module implements the [Compressor] trait by delegating to the segmentors
//! and profilers packages for LLM-based conversation compression.

use std::collections::HashMap;
use std::sync::Arc;

use async_trait::async_trait;

use giztoy_genx::segmentors::{self, SegmentorMux, SegmentorResult};
use giztoy_genx::profilers::{self, ProfilerMux, ProfilerResult};

use crate::error::MemoryError;
use crate::types::{CompressResult, Compressor, EntityInput, EntityUpdate, Message, RelationInput, SegmentInput};

/// LLMCompressorConfig configures an [LLMCompressor].
#[derive(Clone)]
pub struct LLMCompressorConfig {
    /// Segmentor is the pattern of the registered segmentor to use
    /// (e.g., "seg/qwen-flash"). Required. Must be registered in
    /// the provided seg_mux.
    pub segmentor: String,

    /// Profiler is the pattern of the registered profiler to use
    /// (e.g., "prof/qwen-flash"). Optional. If empty, profiling is
    /// skipped and only segmentor output is used.
    pub profiler: Option<String>,

    /// Schema provides entity type hints to guide extraction.
    /// Optional. If nil, the LLM discovers entity types freely.
    pub schema: Option<segmentors::Schema>,

    /// Profiles holds current entity profiles for the profiler to
    /// reference when proposing updates. Optional.
    /// Keyed by entity label (e.g., "person:小明") → attribute map.
    pub profiles: Option<HashMap<String, HashMap<String, serde_json::Value>>>,

    /// SegmentorMux is the multiplexer used to resolve the segmentor. Required.
    pub seg_mux: Arc<SegmentorMux>,

    /// ProfilerMux is the multiplexer used to resolve the profiler.
    /// Required if profiler is set.
    pub prof_mux: Option<Arc<ProfilerMux>>,
}

impl LLMCompressorConfig {
    /// Validate the configuration.
    pub fn validate(&self) -> Result<(), MemoryError> {
        if self.segmentor.is_empty() {
            return Err(MemoryError::General(
                "memory: LLMCompressorConfig.segmentor is required".into(),
            ));
        }
        if self.profiler.is_some() && self.prof_mux.is_none() {
            return Err(MemoryError::General(
                "memory: LLMCompressorConfig.prof_mux is required when profiler is set".into(),
            ));
        }
        Ok(())
    }
}

impl std::fmt::Debug for LLMCompressorConfig {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("LLMCompressorConfig")
            .field("segmentor", &self.segmentor)
            .field("profiler", &self.profiler)
            .field("schema", &self.schema)
            .field("profiles", &self.profiles)
            .field("seg_mux", &"Arc<SegmentorMux>")
            .field("prof_mux", &if self.prof_mux.is_some() { "Some(Arc<ProfilerMux>)" } else { "None" })
            .finish()
    }
}

/// LLMCompressor implements [Compressor] by delegating to the segmentors and
/// profilers packages for LLM-based conversation compression.
///
/// It calls [segmentors::SegmentorMux::process] to extract entities, relations, and a
/// compressed segment summary. Optionally, it calls [profilers::ProfilerMux::process] to
/// evolve entity schemas and update profiles.
///
/// LLMCompressor is stateless and safe for concurrent use.
#[derive(Clone)]
pub struct LLMCompressor {
    cfg: LLMCompressorConfig,
}

impl LLMCompressor {
    /// Create a new LLM-based compressor.
    pub fn new(cfg: LLMCompressorConfig) -> Result<Self, MemoryError> {
        cfg.validate()?;
        Ok(Self { cfg })
    }

    /// Run the segmentor.
    async fn run_segmentor(&self, input: segmentors::SegmentorInput) -> Result<SegmentorResult, MemoryError> {
        let result = self.cfg.seg_mux.process(&self.cfg.segmentor, input).await;
        
        result.map_err(|e| MemoryError::General(e.to_string()))
    }

    /// Run the profiler.
    async fn run_profiler(&self, input: profilers::ProfilerInput) -> Result<ProfilerResult, MemoryError> {
        let pattern = self.cfg.profiler.as_ref().unwrap();
        let result = if let Some(mux) = &self.cfg.prof_mux {
            mux.process(pattern, input).await
        } else {
            return Err(MemoryError::General("memory: prof_mux is required".into()));
        };
        
        result.map_err(|e| MemoryError::General(e.to_string()))
    }
}

#[async_trait]
impl Compressor for LLMCompressor {
    async fn compress_messages(&self, messages: &[Message]) -> Result<CompressResult, MemoryError> {
        let input = segmentors::SegmentorInput {
            messages: messages_to_strings(messages),
            schema: self.cfg.schema.clone(),
        };
        
        let result = self.run_segmentor(input).await?;

        // Convert segmentor output to CompressResult.
        let summary = result.segment.summary.clone();
        let seg = SegmentInput {
            summary: result.segment.summary,
            keywords: result.segment.keywords,
            labels: result.segment.labels,
        };

        Ok(CompressResult {
            segments: vec![seg],
            summary,
        })
    }

    async fn extract_entities(&self, messages: &[Message]) -> Result<EntityUpdate, MemoryError> {
        let input = segmentors::SegmentorInput {
            messages: messages_to_strings(messages),
            schema: self.cfg.schema.clone(),
        };

        let result = self.run_segmentor(input).await?;

        let mut update = convert_segmentor_result(&result);

        // Run profiler if configured.
        if self.cfg.profiler.is_some() {
            let prof_input = profilers::ProfilerInput {
                messages: messages_to_strings(messages),
                extracted: result.clone(),
                schema: self.cfg.schema.clone(),
                profiles: self.cfg.profiles.clone(),
            };
            if let Ok(prof_result) = self.run_profiler(prof_input).await {
                merge_profiler_result(&mut update, &prof_result);
            }
            // Profiler failure is non-fatal - we still have the segmentor result.
        }

        Ok(update)
    }

    async fn compact_segments(&self, summaries: &[String]) -> Result<CompressResult, MemoryError> {
        let input = segmentors::SegmentorInput {
            messages: summaries.to_vec(),
            schema: self.cfg.schema.clone(),
        };

        let result = self.run_segmentor(input).await?;

        let summary = result.segment.summary.clone();
        let seg = SegmentInput {
            summary: result.segment.summary,
            keywords: result.segment.keywords,
            labels: result.segment.labels,
        };

        Ok(CompressResult {
            segments: vec![seg],
            summary,
        })
    }
}

/// Convert segmentor entities and relations into a memory.EntityUpdate.
fn convert_segmentor_result(result: &SegmentorResult) -> EntityUpdate {
    let mut entities = Vec::new();
    for e in &result.entities {
        entities.push(EntityInput {
            label: e.label.clone(),
            attrs: e.attrs.clone(),
        });
    }

    let mut relations = Vec::new();
    for r in &result.relations {
        relations.push(RelationInput {
            from: r.from.clone(),
            to: r.to.clone(),
            rel_type: r.rel_type.clone(),
        });
    }

    EntityUpdate { entities, relations }
}

/// Merge profiler output into an existing EntityUpdate.
fn merge_profiler_result(update: &mut EntityUpdate, prof_result: &ProfilerResult) {
    // Merge profile updates as entity attrs.
    for (label, attrs) in &prof_result.profile_updates {
        let existing = update.entities.iter_mut().find(|e| e.label == *label);
        if let Some(existing_entity) = existing {
            for (k, v) in attrs {
                existing_entity.attrs.insert(k.clone(), v.clone());
            }
        } else {
            update.entities.push(EntityInput {
                label: label.clone(),
                attrs: attrs.clone(),
            });
        }
    }

    // Append additional relations from profiler.
    for r in &prof_result.relations {
        update.relations.push(RelationInput {
            from: r.from.clone(),
            to: r.to.clone(),
            rel_type: r.rel_type.clone(),
        });
    }
}

/// Convert memory.Message slice to the plain string format expected by segmentors/profilers:
/// "role: content" or "role(name): content".
pub fn messages_to_strings(messages: &[Message]) -> Vec<String> {
    messages
        .iter()
        .filter(|m| !m.content.is_empty())
        .map(|m| {
            let mut s = String::new();
            s.push_str(match m.role {
                crate::types::Role::User => "user",
                crate::types::Role::Model => "model",
                crate::types::Role::Tool => "tool",
            });
            if !m.name.is_empty() {
                s.push('(');
                s.push_str(&m.name);
                s.push(')');
            }
            s.push_str(": ");
            s.push_str(&m.content);
            s
        })
        .collect()
}
